src/kakarot/instructions/system_operations.cairo

```diff
@@ -212,19 +212,31 @@ namespace SystemOperations {
         // 0 - size: byte size to copy
         // 1 - offset: byte offset in the memory in bytes
         let (stack, popped) = Stack.pop_n(self=stack, n=2);
-        // TODO: implement loading of the revert reason based on size value,
-        // currently limited by short string size
+
         let size = popped[0];
         let offset = popped[1];
 
         // Load revert reason from offset
-        let (memory, revert_reason_uint256, gas_cost) = Memory.load(memory, offset.low);
-        local revert_reason = revert_reason_uint256.low;
+        let (revert_reason_bytes: felt*) = alloc();
+        let (memory, gas_cost) = Memory.load_n(memory, size.low, revert_reason_bytes, offset.low);
+
+        // revert with loaded revert reason short string: 31 bytes of the last word
+        let reason_is_single_word = is_le(size.low, 32);
+        if (reason_is_single_word != FALSE) {
+            tempvar initial_byte: felt* = revert_reason_bytes;
+            tempvar actual_size = size.low;
+        } else {
+            tempvar byte_shift = size.low - 32;
+            tempvar initial_byte: felt* = revert_reason_bytes + byte_shift;
```

Comment:
does this mean that we take the 31 last bytes and shift to the last 32 bytes? Does this mean that essentially the very last byte is ignored?

Reply:
yes I've assumed that the last word (`bytes32`)  is more relevant than the first one (because of how `string` encoding works) but eventually the revert_reason is a `bytes31` (a `felt`) so we can keep it all, and consequently I've decided to use the first 31 bytes of the last word, or in other words we drop the very last byte.

I don't know tbh if it's possible to have a size that is not a multiple of `32`.